File: modules/virtual-people-core-serving/src/main/kotlin/org/wfanet/virtualpeople/core/model/MultiplicityImpl.kt
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.core.model

import com.google.common.hash.Hashing
import com.google.protobuf.Descriptors.FieldDescriptor
import com.google.protobuf.Descriptors.FieldDescriptor.Type
import java.nio.charset.StandardCharsets
import kotlin.math.floor
import org.wfanet.virtualpeople.common.LabelerEvent
import org.wfanet.virtualpeople.common.LabelerEventOrBuilder
import org.wfanet.virtualpeople.common.Multiplicity
import org.wfanet.virtualpeople.common.Multiplicity.MultiplicityRefCase
import org.wfanet.virtualpeople.common.fieldfilter.utils.getFieldFromProto
import org.wfanet.virtualpeople.common.fieldfilter.utils.getValueFromProto

class MultiplicityImpl
/**
 * @param multiplicityExtractor extractor for expected multiplicity. Either a [Double] or a
 * [MultiplicityFromField].
 * @param capAtMax whether to cap multiplicity at [maxValue]
 * @param maxValue when expected multiplicity > [maxValue], cap at [maxValue] if [capAtMax] is YES,
 * else throw an error.
 * @param personIndexField the field to set person index in.
 * @param randomSeed the random seed. It is used to compute multiplicity for a given event and
 * compute fingerprint for cloned event
 */
private constructor(
  private val multiplicityExtractor: Any,
  private val capAtMax: CapMultiplicityAtMax,
  private val maxValue: Double,
  private val personIndexField: List<FieldDescriptor>,
  private val randomSeed: String
) {

  /**
   * Computes multiplicity for [eventOrBuilder].
   * 1. Extracts the expected_multiplicity. Throws an error if [capAtMax] = NO and
   * expected_multiplicity > [maxValue].
   * 2. Pseudo-randomly generates an integer value that is either floor(expected_multiplicity) or
   * floor(expected_multiplicity) + 1, with expectation = expected_multiplicity. For example, with
   * expected_multiplicity = 1.4, this returns either 1 or 2, with 60% and 40% probabilities,
   * respectively. This always returns the same result for the same event.
   */
  fun computeEventMultiplicity(eventOrBuilder: LabelerEventOrBuilder): Int {
    var expectedMultiplicity =
      when (multiplicityExtractor) {
        is Double -> multiplicityExtractor
        is MultiplicityFromField -> {
          if (multiplicityExtractor.fieldDescriptors.isEmpty()) {
            error("Extractor has invalid field_descriptor.")
          }
          multiplicityExtractor.getValueFunction(
            eventOrBuilder,
            multiplicityExtractor.fieldDescriptors
          )
        }
        else -> error("Invalid multiplicity extractor.")
      }
    if (expectedMultiplicity > maxValue) {
      if (capAtMax == CapMultiplicityAtMax.YES) {
        expectedMultiplicity = maxValue
      } else {
        error(
          "Expected multiplicity = $expectedMultiplicity," +
            ", which exceeds the specified max value = $maxValue"
        )
      }
    }
    if (expectedMultiplicity < 0) {
      error("Expected multiplicity = $expectedMultiplicity, but multiplicity must >= 0.")
    }

    /**
     * The seed uses the string representation of actingFingerprint as an unsigned 64-bit integer
     */
    val eventSeed =
      Hashing.farmHashFingerprint64()
        .hashString(
          "$randomSeed${eventOrBuilder.actingFingerprint.toULong()}",
          StandardCharsets.UTF_8
        )
        .asLong()
        .toULong()
    return computeBimodalInteger(expectedMultiplicity, eventSeed)
  }

  /** Gets the person_index field descriptor. */
  fun personIndexFieldDescriptor(): List<FieldDescriptor> {
    return personIndexField
  }

  /**
   * Gets fingerprint for [index] using [input] and [randomSeed].
   *
   * Returns [input] as is for [index] = 0.
   */
  fun getFingerprintForIndex(input: ULong, index: Int): ULong {
    if (index == 0) {
      return input
    }
    return Hashing.farmHashFingerprint64()
      .hashString("$randomSeed-clone-$index-$input", StandardCharsets.UTF_8)
      .asLong()
      .toULong()
  }

  companion object {
    private inline fun <reified T> extractValue(
      eventOrBuilder: LabelerEventOrBuilder,
      source: List<FieldDescriptor>
    ): T {
      val fieldValue = getValueFromProto<T>(eventOrBuilder, source)
      if (fieldValue.isSet) {
        return fieldValue.value
      } else {
        error("The multiplicity field is not set.")
      }
    }

    /**
     * Returns an integer that is either N or N + 1, where N is the integral part of [expectation].
     *
     * The choice is made by comparing [seed] to the fractional part of [expectation]. This is used
     * to compute multiplicity, and [expectation] should be much lower than int max.
     */
    private fun computeBimodalInteger(expectation: Double, seed: ULong): Int {
      val integralPart: Int = floor(expectation).toInt()
      val fractionalPart: Double = expectation - integralPart
      val threshold: ULong = (fractionalPart * ULong.MAX_VALUE.toDouble()).toULong()
      return if (seed < threshold) integralPart + 1 else integralPart
    }

    /** Gets a function to extract multiplicity value from a field. */
    private fun getExtractMultiplicityFunction(
      type: Type
    ): (LabelerEventOrBuilder, List<FieldDescriptor>) -> Double {
      return when (type) {
        Type.INT32 -> { a, b -> extractValue<Int>(a, b).toDouble() }
        Type.UINT32 -> { a, b -> extractValue<UInt>(a, b).toDouble() }
        Type.INT64 -> { a, b -> extractValue<Long>(a, b).toDouble() }
        Type.UINT64 -> { a, b -> extractValue<ULong>(a, b).toDouble() }
        Type.FLOAT -> { a, b -> extractValue<Float>(a, b).toDouble() }
        Type.DOUBLE -> { a, b -> extractValue(a, b) }
        else -> error("Unsupported field type for multiplicity: $type")
      }
    }

    private fun isIntegerFieldType(type: Type): Boolean {
      return type in listOf(Type.INT32, Type.UINT32, Type.INT64, Type.UINT64)
    }

    /**
     * Always use build to get an [MultiplicityImpl] object. Users should not call the constructor
     * directly.
     *
     * Throws an error if any of the following happens:
     * 1. [config].multiplicity_ref is not set.
     * 2. [config].expected_multiplicity_field is set, but is not a valid field, or the field type
     * is not one of int32/int64/uint32/uint64/float/double.
     * 3. [config].person_index_field is not set, or is not a valid field, or the field type is not
     * one of int32/int64/uint32/uint64.
     * 4. [config].max_value is not set.
     * 5. [config].cap_at_max is not set.
     * 6. [config].random_seed is not set.
     */
    fun build(config: Multiplicity): MultiplicityImpl {
      if (!config.hasPersonIndexField()) {
        error("Multiplicity must set person_index_field. $config")
      }
      if (!config.hasMaxValue()) {
        error("Multiplicity must set max_value. $config")
      }
      if (!config.hasCapAtMax()) {
        error("Multiplicity must set cap_at_max. $config")
      }
      if (!config.hasRandomSeed()) {
        error("Multiplicity must set random_seed. $config")
      }

      /** Only [Double] or [MultiplicityFromField] is valid type for [multiplicityExtractor]. */
      val multiplicityExtractor: Any =
        @Suppress("WHEN_ENUM_CAN_BE_NULL_IN_JAVA") // Proto enum fields are never null.
        when (config.multiplicityRefCase) {
          MultiplicityRefCase.EXPECTED_MULTIPLICITY -> config.expectedMultiplicity
          MultiplicityRefCase.EXPECTED_MULTIPLICITY_FIELD -> {
            val fieldDescriptors =
              getFieldFromProto(LabelerEvent.getDescriptor(), config.expectedMultiplicityField)
            MultiplicityFromField(
              fieldDescriptors,
              getExtractMultiplicityFunction(fieldDescriptors.last().type)
            )
          }
          MultiplicityRefCase.MULTIPLICITYREF_NOT_SET ->
            error("Multiplicity must set multiplicity_ref. $config")
        }

      val personIndexField =
        getFieldFromProto(LabelerEvent.getDescriptor(), config.personIndexField)
      if (!isIntegerFieldType(personIndexField.last().type)) {
        error("Invalid type for person_index_field. $config")
      }
      val capAtMax = if (config.capAtMax) CapMultiplicityAtMax.YES else CapMultiplicityAtMax.NO

      return MultiplicityImpl(
        multiplicityExtractor,
        capAtMax,
        config.maxValue,
        personIndexField,
        config.randomSeed
      )
    }
  }
}

enum class CapMultiplicityAtMax {
  NO,
  YES
}

/** Extracts multiplicity value from a field. */
data class MultiplicityFromField(
  /** The field descriptor. */
  val fieldDescriptors: List<FieldDescriptor>,
  /** The function to extract the multiplicity value. */
  val getValueFunction: (LabelerEventOrBuilder, List<FieldDescriptor>) -> Double
)

File: modules/virtual-people-core-serving/src/main/cc/wfa/virtual_people/core/model/multiplicity_impl.cc
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "wfa/virtual_people/core/model/multiplicity_impl.h"

#include <limits>
#include <memory>
#include <utility>
#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/types/variant.h"
#include "common_cpp/macros/macros.h"
#include "src/farmhash.h"
#include "wfa/virtual_people/common/field_filter/utils/field_util.h"
#include "wfa/virtual_people/common/model.pb.h"

namespace wfa_virtual_people {

// Returns an integer that is either N or N + 1, where N is the integral
// part of @expectation.
// The choice is made by comparing @seed to the fractional part of @expectation.
// This is used to compute multiplicity, and @expectation should be much lower
// than int max.
absl::StatusOr<int> ComputeBimodalInteger(double expectation, uint64_t seed) {
  int integral_part = static_cast<int>(floor(expectation));
  double fractional_part = expectation - integral_part;
  uint64_t threshold = static_cast<uint64_t>(
      fractional_part *
      static_cast<double>(std::numeric_limits<uint64_t>::max()));
  if (seed < threshold) {
    return integral_part + 1;
  } else {
    return integral_part;
  }
}

template <typename ValueType>
using IsIntegerFloatType =
    absl::disjunction<IsIntegerType<ValueType>, std::is_same<ValueType, float>,
                      std::is_same<ValueType, double>>;

template <typename ValueType>
using EnableIfIntegerFloatType =
    absl::enable_if_t<IsIntegerFloatType<ValueType>::value, bool>;

// Extracts a double value from the @source field in @event.
template <typename ValueType, EnableIfIntegerFloatType<ValueType> = true>
absl::StatusOr<double> ExtractDoubleValue(
    const LabelerEvent& event,
    const std::vector<const google::protobuf::FieldDescriptor*>& source) {
  ProtoFieldValue<ValueType> field_value =
      GetValueFromProto<ValueType>(event, source);
  if (field_value.is_set) {
    return field_value.value;
  }

  return absl::InvalidArgumentError("The multiplicity field is not set.");
}

// Gets a function to extract multiplicity value from a field.
absl::StatusOr<std::function<absl::StatusOr<double>(
    const LabelerEvent&,
    const std::vector<const google::protobuf::FieldDescriptor*>&)>>
GetExtractMultiplicityFunction(
    const google::protobuf::FieldDescriptor::CppType cpp_type) {
  switch (cpp_type) {
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_INT32:
      return ExtractDoubleValue<int32_t>;
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_INT64:
      return ExtractDoubleValue<int64_t>;
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_UINT32:
      return ExtractDoubleValue<uint32_t>;
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_UINT64:
      return ExtractDoubleValue<uint64_t>;
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_FLOAT:
      return ExtractDoubleValue<float>;
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_DOUBLE:
      return ExtractDoubleValue<double>;
    default:
      return absl::InvalidArgumentError(
          "Unsupported field type for multiplicity.");
  }
}

bool IsIntegerFieldType(
    const google::protobuf::FieldDescriptor::CppType cpp_type) {
  switch (cpp_type) {
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_INT32:
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_INT64:
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_UINT32:
    case google::protobuf::FieldDescriptor::CppType::CPPTYPE_UINT64:
      return true;
    default:
      return false;
  }

  return false;
}

absl::StatusOr<std::unique_ptr<MultiplicityImpl>> MultiplicityImpl::Build(
    const Multiplicity& config) {
  MultiplicityExtractor multiplicity_extractor;
  switch (config.multiplicity_ref_case()) {
    case Multiplicity::kExpectedMultiplicity: {
      multiplicity_extractor = config.expected_multiplicity();
      break;
    }
    case Multiplicity::kExpectedMultiplicityField: {
      ASSIGN_OR_RETURN(std::vector<const google::protobuf::FieldDescriptor*>
                           field_descriptor,
                       GetFieldFromProto(LabelerEvent().GetDescriptor(),
                                         config.expected_multiplicity_field()));
      MultiplicityFromField from_field;
      from_field.field_descriptor = std::move(field_descriptor);
      ASSIGN_OR_RETURN(from_field.get_value_function,
                       GetExtractMultiplicityFunction(
                           from_field.field_descriptor.back()->cpp_type()));
      multiplicity_extractor = std::move(from_field);
      break;
    }
    default:
      // multiplicity_ref is not set.
      return absl::InvalidArgumentError(absl::StrCat(
          "Multiplicity must set multiplicity_ref.", config.DebugString()));
  }

  if (!config.has_person_index_field()) {
    return absl::InvalidArgumentError(absl::StrCat(
        "Multiplicity must set person_index_field.", config.DebugString()));
  }
  ASSIGN_OR_RETURN(
      std::vector<const google::protobuf::FieldDescriptor*> person_index_field,
      GetFieldFromProto(LabelerEvent().GetDescriptor(),
                        config.person_index_field()));
  if (!IsIntegerFieldType(person_index_field.back()->cpp_type())) {
    return absl::InvalidArgumentError(absl::StrCat(
        "Invalid type for person_index_field.", config.DebugString()));
  }

  if (!config.has_max_value()) {
    return absl::InvalidArgumentError(
        absl::StrCat("Multiplicity must set max_value.", config.DebugString()));
  }
  if (!config.has_cap_at_max()) {
    return absl::InvalidArgumentError(absl::StrCat(
        "Multiplicity must set cap_at_max.", config.DebugString()));
  }
  if (!config.has_random_seed()) {
    return absl::InvalidArgumentError(absl::StrCat(
        "Multiplicity must set random_seed.", config.DebugString()));
  }

  CapMultiplicityAtMax cap_at_max = config.cap_at_max()
                                        ? CapMultiplicityAtMax::kYes
                                        : CapMultiplicityAtMax::kNo;
  return absl::make_unique<MultiplicityImpl>(
      multiplicity_extractor, cap_at_max, config.max_value(),
      std::move(person_index_field), config.random_seed());
}

MultiplicityImpl::MultiplicityImpl(
    MultiplicityExtractor& multiplicity_extractor,
    CapMultiplicityAtMax cap_at_max, double max_value,
    std::vector<const google::protobuf::FieldDescriptor*>&& person_index_field,
    absl::string_view random_seed)
    : multiplicity_extractor_(std::move(multiplicity_extractor)),
      cap_at_max_(cap_at_max),
      max_value_(max_value),
      person_index_field_(std::move(person_index_field)),
      random_seed_(random_seed) {}

absl::StatusOr<int> MultiplicityImpl::ComputeEventMultiplicity(
    const LabelerEvent& event) const {
  double expected_multiplicity = 0.0;
  if (const double* explicit_value =
          std::get_if<double>(&multiplicity_extractor_)) {
    expected_multiplicity = *explicit_value;
  } else if (const MultiplicityFromField* from_field =
                 std::get_if<MultiplicityFromField>(&multiplicity_extractor_)) {
    if (!from_field->get_value_function) {
      return absl::InternalError("Extractor has NULL get_value_function.");
    }

    if (from_field->field_descriptor.empty()) {
      return absl::InternalError("Extractor has invalid field_descriptor.");
    }

    ASSIGN_OR_RETURN(
        expected_multiplicity,
        from_field->get_value_function(event, from_field->field_descriptor));
  } else {
    return absl::InternalError("Invalid multiplicity extractor.");
  }

  if (expected_multiplicity > max_value_) {
    if (cap_at_max_ == CapMultiplicityAtMax::kYes) {
      expected_multiplicity = max_value_;
    } else {
      return absl::OutOfRangeError(absl::StrCat(
          "Expected multiplicity = ", expected_multiplicity,
          ", which exceeds the specified max value = ", max_value_));
    }
  }

  if (expected_multiplicity < 0) {
    return absl::OutOfRangeError(
        absl::StrCat("Expected multiplicity = ", expected_multiplicity,
                     ", but multiplicity must >= 0."));
  }

  uint64_t event_seed = util::Fingerprint64(
      absl::StrCat(random_seed_, event.acting_fingerprint()));
  return ComputeBimodalInteger(expected_multiplicity, event_seed);
}

const std::vector<const google::protobuf::FieldDescriptor*>&
MultiplicityImpl::PersonIndexFieldDescriptor() const {
  return person_index_field_;
}

uint64_t MultiplicityImpl::GetFingerprintForIndex(uint64_t input,
                                                  int index) const {
  if (index == 0) {
    return input;
  }

  return util::Fingerprint64(
      absl::StrCat(random_seed_, "-clone-", index, "-", input));
}

}  // namespace wfa_virtual_people
